File: internal/sdk/go/mindclade/policies.go
package mindclade

import (
	"context"
	"strings"

	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/timestamppb"

	commonv1 "github.com/mindclade/mindclade/protocols/generated/go/common/v1"
	internalpolicyv1 "github.com/mindclade/mindclade/protocols/generated/go/internalrpc/policy/v1"
	jobv1 "github.com/mindclade/mindclade/protocols/generated/go/job/v1"
	policyv1 "github.com/mindclade/mindclade/protocols/generated/go/policy/v1"
)

// PolicyService is the private generated-type-only policy lifecycle and
// authorization facade. Authenticated SDK configuration, never caller-owned
// request fields, is authoritative for tenant, project, and principal scope.
type PolicyService struct {
	client    *Client
	transport internalpolicyv1.PolicyServiceClient
}

func (service *PolicyService) Evaluate(ctx context.Context, request *internalpolicyv1.EvaluateAuthorizationRequest, options ...RequestOption) (*policyv1.AuthorizationDecision, error) {
	materialized := cloneGenerated(request)
	if materialized == nil || strings.TrimSpace(materialized.GetAction()) == "" || materialized.GetResource() == nil || !validSHA256Digest(materialized.GetIntentDigest()) {
		return nil, invalidArgument("authorization evaluation requires an action, resource, and sha256 intent digest")
	}
	if !normalizePolicyResource(service.client.config, materialized.Resource) {
		return nil, invalidArgument("authorization resource must be in the configured project")
	}
	materialized.TenantId = service.client.config.TenantID
	materialized.ProjectId = service.client.config.ProjectID
	materialized.PrincipalRef = service.client.config.PrincipalID
	key := materialized.GetContext().GetIdempotencyKey()
	materialized.Context = nil
	callContext, metadata, cancel, err := service.client.mutationContext(ctx, key, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	deadline, _ := callContext.Deadline()
	materialized.Deadline = timestamppb.New(deadline)
	digest, err := deterministicDigest(materialized)
	if err != nil {
		return nil, err
	}
	materialized.Context = commandContext(service.client.config, callContext, metadata, digest)
	response, err := service.transport.EvaluateAuthorization(callContext, materialized)
	if err != nil {
		return nil, normalizeError(err)
	}
	if response.GetDecision() == nil {
		return nil, protocolDataLoss("EvaluateAuthorization response omitted its decision")
	}
	return cloneGenerated(response.GetDecision()), nil
}

func (service *PolicyService) Create(ctx context.Context, request *internalpolicyv1.CreateUsePolicyRequest, options ...RequestOption) (*jobv1.Operation, error) {
	materialized := cloneGenerated(request)
	if materialized == nil || materialized.GetUsePolicy() == nil || !validResourceIdentifier(materialized.GetUsePolicyId()) {
		return nil, invalidArgument("policy create requires a generated policy and valid policy ID")
	}
	parent := projectName(service.client.config.TenantID, service.client.config.ProjectID)
	if materialized.GetParent() != "" && materialized.GetParent() != parent {
		return nil, invalidArgument("policy parent must match the configured project")
	}
	materialized.Parent = parent
	callContext, cancel, err := service.prepareMutation(ctx, materialized, materialized.GetContext(), func(value *commonv1.CommandContext) { materialized.Context = value }, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	response, rpcErr := service.transport.CreateUsePolicy(callContext, materialized)
	return operationResponse(response.GetOperation(), rpcErr, "CreateUsePolicy")
}

func (service *PolicyService) Update(ctx context.Context, request *internalpolicyv1.UpdateUsePolicyRequest, options ...RequestOption) (*jobv1.Operation, error) {
	materialized := cloneGenerated(request)
	if materialized == nil || materialized.GetUsePolicy() == nil || !scopedResourceName(service.client.config, materialized.GetUsePolicy().GetName(), "usePolicies") || materialized.GetUpdateMask() == nil || strings.TrimSpace(materialized.GetEtag()) == "" {
		return nil, invalidArgument("policy update requires a scoped policy, field mask, and etag")
	}
	callContext, cancel, err := service.prepareMutation(ctx, materialized, materialized.GetContext(), func(value *commonv1.CommandContext) { materialized.Context = value }, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	response, rpcErr := service.transport.UpdateUsePolicy(callContext, materialized)
	return operationResponse(response.GetOperation(), rpcErr, "UpdateUsePolicy")
}

func (service *PolicyService) Get(ctx context.Context, name, ifNoneMatch string, options ...RequestOption) (*policyv1.UsePolicy, error) {
	if !scopedResourceName(service.client.config, name, "usePolicies") {
		return nil, invalidArgument("policy name must be in the configured project")
	}
	callContext, _, cancel, err := service.client.context(ctx, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	response, err := service.transport.GetUsePolicy(callContext, &internalpolicyv1.GetUsePolicyRequest{Name: name, IfNoneMatch: strings.TrimSpace(ifNoneMatch)})
	if err != nil {
		return nil, normalizeError(err)
	}
	if response.GetUsePolicy() == nil {
		return nil, protocolDataLoss("GetUsePolicy response omitted its policy")
	}
	return cloneGenerated(response.GetUsePolicy()), nil
}

// UsePolicyPage is one bounded list response plus cursor-scheme traversal. The
// embedded generated response remains the authoritative model; the wrapper
// adds only the opaque-cursor mechanics.
type UsePolicyPage struct {
	*internalpolicyv1.ListUsePoliciesResponse
	pageBase[*policyv1.UsePolicy, *UsePolicyPage]
}

// Items returns this page's use policies without traversing any further page.
func (page *UsePolicyPage) Items() []*policyv1.UsePolicy { return page.GetUsePolicies() }

func (service *PolicyService) List(ctx context.Context, request *internalpolicyv1.ListUsePoliciesRequest, options ...RequestOption) (*UsePolicyPage, error) {
	materialized := cloneGenerated(request)
	if materialized == nil {
		materialized = &internalpolicyv1.ListUsePoliciesRequest{}
	}
	parent := projectName(service.client.config.TenantID, service.client.config.ProjectID)
	if materialized.GetParent() != "" && materialized.GetParent() != parent {
		return nil, invalidArgument("policy list parent must match the configured project")
	}
	if materialized.GetPage().GetPageSize() > 1000 {
		return nil, invalidArgument("policy page size cannot exceed 1000")
	}
	materialized.Parent = parent
	callContext, _, cancel, err := service.client.context(ctx, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	response, err := service.transport.ListUsePolicies(callContext, materialized)
	if err != nil {
		return nil, normalizeError(err)
	}
	detached := cloneGenerated(response)
	page := &UsePolicyPage{ListUsePoliciesResponse: detached}
	page.pageBase = newPage[*policyv1.UsePolicy](page, detached.GetPage(), paginationLimitsFrom(options), func(ctx context.Context, token string) (*UsePolicyPage, error) {
		successor := cloneGenerated(materialized)
		successor.Page = pageRequestWithToken(materialized.GetPage(), token)
		return service.List(ctx, successor, options...)
	})
	return page, nil
}

func (service *PolicyService) Activate(ctx context.Context, name, etag string, options ...RequestOption) (*jobv1.Operation, error) {
	if !scopedResourceName(service.client.config, name, "usePolicies") || strings.TrimSpace(etag) == "" {
		return nil, invalidArgument("policy activation requires a scoped name and etag")
	}
	request := &internalpolicyv1.ActivateUsePolicyRequest{Name: name, Etag: strings.TrimSpace(etag)}
	callContext, cancel, err := service.prepareMutation(ctx, request, request.GetContext(), func(value *commonv1.CommandContext) { request.Context = value }, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	response, rpcErr := service.transport.ActivateUsePolicy(callContext, request)
	return operationResponse(response.GetOperation(), rpcErr, "ActivateUsePolicy")
}

func (service *PolicyService) Revoke(ctx context.Context, name, etag, reasonCode string, options ...RequestOption) (*jobv1.Operation, error) {
	if !scopedResourceName(service.client.config, name, "usePolicies") || strings.TrimSpace(etag) == "" || strings.TrimSpace(reasonCode) == "" {
		return nil, invalidArgument("policy revocation requires a scoped name, etag, and reason code")
	}
	request := &internalpolicyv1.RevokeUsePolicyRequest{Name: name, Etag: strings.TrimSpace(etag), ReasonCode: strings.TrimSpace(reasonCode)}
	callContext, cancel, err := service.prepareMutation(ctx, request, request.GetContext(), func(value *commonv1.CommandContext) { request.Context = value }, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	response, rpcErr := service.transport.RevokeUsePolicy(callContext, request)
	return operationResponse(response.GetOperation(), rpcErr, "RevokeUsePolicy")
}

func (service *PolicyService) ResolveSnapshot(ctx context.Context, name string, effectiveTime *timestamppb.Timestamp, options ...RequestOption) (*policyv1.PolicyReference, error) {
	if !scopedResourceName(service.client.config, name, "usePolicies") || effectiveTime == nil || effectiveTime.CheckValid() != nil {
		return nil, invalidArgument("snapshot resolution requires a scoped policy and valid effective time")
	}
	callContext, _, cancel, err := service.client.context(ctx, options...)
	if err != nil {
		return nil, err
	}
	defer cancel()
	response, err := service.transport.ResolvePolicySnapshot(callContext, &internalpolicyv1.ResolvePolicySnapshotRequest{Name: name, EffectiveTime: cloneGenerated(effectiveTime)})
	if err != nil {
		return nil, normalizeError(err)
	}
	if response.GetPolicySnapshot() == nil {
		return nil, protocolDataLoss("ResolvePolicySnapshot response omitted its snapshot")
	}
	return cloneGenerated(response.GetPolicySnapshot()), nil
}

func (service *PolicyService) prepareMutation(ctx context.Context, request proto.Message, existing *commonv1.CommandContext, assign func(*commonv1.CommandContext), options ...RequestOption) (context.Context, context.CancelFunc, error) {
	key := existing.GetIdempotencyKey()
	assign(nil)
	callContext, metadata, cancel, err := service.client.mutationContext(ctx, key, options...)
	if err != nil {
		return nil, nil, err
	}
	digest, err := deterministicDigest(request)
	if err != nil {
		cancel()
		return nil, nil, err
	}
	assign(commandContext(service.client.config, callContext, metadata, digest))
	return callContext, cancel, nil
}

func normalizePolicyResource(config Config, resource *commonv1.ResourceRef) bool {
	if resource == nil || (resource.GetTenantId() != "" && resource.GetTenantId() != config.TenantID) || (resource.GetProjectId() != "" && resource.GetProjectId() != config.ProjectID) {
		return false
	}
	parent := projectName(config.TenantID, config.ProjectID)
	if resource.GetName() != parent && !strings.HasPrefix(resource.GetName(), parent+"/") {
		return false
	}
	resource.TenantId = config.TenantID
	resource.ProjectId = config.ProjectID
	return true
}

func validSHA256Digest(value string) bool {
	if len(value) != 71 || !strings.HasPrefix(value, "sha256:") {
		return false
	}
	for _, character := range value[7:] {
		if (character < '0' || character > '9') && (character < 'a' || character > 'f') {
			return false
		}
	}
	return true
}

File: sdks/python/mindclade_internal_sdk/policies.py
"""Private Policy lifecycle and fail-closed authorization façade."""

from __future__ import annotations

import copy
import re
from dataclasses import replace
from typing import cast

from google.protobuf.timestamp_pb2 import Timestamp
from mindclade.internal.policy.v1 import policy_service_pb2
from mindclade.operation.v1 import operation_pb2
from mindclade.policy.v1 import authorization_decision_pb2, policy_reference_pb2, use_policy_pb2

from ._invocation import AsyncInvoker, SyncInvoker, canonical_digest, command_context
from ._validation import required_response_message, required_text
from .calls import CallOptions, PreparedCall, prepare_call
from .transport import (
    ACTIVATE_USE_POLICY,
    CREATE_USE_POLICY,
    EVALUATE_AUTHORIZATION,
    GET_USE_POLICY,
    LIST_USE_POLICIES,
    RESOLVE_POLICY_SNAPSHOT,
    REVOKE_USE_POLICY,
    UPDATE_USE_POLICY,
)

_DIGEST = re.compile(r"sha256:[0-9a-f]{64}\Z")


def _project(invoker: SyncInvoker | AsyncInvoker) -> str:
    return invoker.config.project_parent


def _policy_name(invoker: SyncInvoker | AsyncInvoker, value: str) -> str:
    name = required_text("policy name", value)
    prefix = f"{_project(invoker)}/usePolicies/"
    suffix = name.removeprefix(prefix)
    if not name.startswith(prefix) or not suffix or "/" in suffix:
        raise ValueError("policy name must be scoped to the configured project")
    return name


def _mutation_options(key: str, options: CallOptions | None) -> CallOptions | None:
    if not key or (options is not None and options.idempotency_key is not None):
        return options
    return replace(options, idempotency_key=key) if options else CallOptions(idempotency_key=key)


def _prepare_mutation[
    RequestT: (
        policy_service_pb2.CreateUsePolicyRequest,
        policy_service_pb2.UpdateUsePolicyRequest,
        policy_service_pb2.ActivateUsePolicyRequest,
        policy_service_pb2.RevokeUsePolicyRequest,
    )
](
    invoker: SyncInvoker | AsyncInvoker,
    request: RequestT,
    options: CallOptions | None,
) -> tuple[RequestT, PreparedCall]:
    materialized = copy.deepcopy(request)
    key = materialized.context.idempotency_key if materialized.HasField("context") else ""
    materialized.ClearField("context")
    call = prepare_call(
        _mutation_options(key, options),
        default_timeout=invoker.config.default_timeout,
        require_idempotency=True,
    )
    materialized.context.CopyFrom(
        command_context(invoker.config, call, request_digest=canonical_digest(materialized))
    )
    return materialized, call


def _operation(response: object, label: str) -> operation_pb2.Operation:
    operation = required_response_message(
        cast(policy_service_pb2.CreateUsePolicyResponse, response),
        "operation",
        operation_pb2.Operation,
        label=label,
    )
    required_text("operation id", operation.operation_id)
    return operation


def _prepare_evaluation(
    invoker: SyncInvoker | AsyncInvoker,
    request: policy_service_pb2.EvaluateAuthorizationRequest,
    options: CallOptions | None,
) -> tuple[policy_service_pb2.EvaluateAuthorizationRequest, PreparedCall]:
    materialized = copy.deepcopy(request)
    if not materialized.action.strip() or not materialized.HasField("resource"):
        raise ValueError("authorization evaluation requires an action and resource")
    if not _DIGEST.fullmatch(materialized.intent_digest):
        raise ValueError("authorization intent_digest must be lowercase sha256")
    project = _project(invoker)
    if materialized.resource.name != project and not materialized.resource.name.startswith(
        f"{project}/"
    ):
        raise ValueError("authorization resource is outside the configured project")
    config = invoker.config
    if materialized.resource.tenant_id not in ("", config.tenant_id) or (
        materialized.resource.project_id not in ("", config.project_id)
    ):
        raise ValueError("authorization resource scope conflicts with client identity")
    materialized.tenant_id = config.tenant_id
    materialized.project_id = config.project_id
    materialized.principal_ref = config.principal_id
    materialized.resource.tenant_id = config.tenant_id
    materialized.resource.project_id = config.project_id
    key = materialized.context.idempotency_key if materialized.HasField("context") else ""
    materialized.ClearField("context")
    call = prepare_call(
        _mutation_options(key, options),
        default_timeout=config.default_timeout,
        require_idempotency=True,
    )
    authoritative = command_context(config, call, request_digest="")
    materialized.deadline.CopyFrom(authoritative.deadline)
    authoritative.canonical_request_digest = canonical_digest(materialized)
    materialized.context.CopyFrom(authoritative)
    return materialized, call


class Policies:
    """Synchronous generated-type-only Policy API."""

    def __init__(self, invoker: SyncInvoker) -> None:
        self._invoker = invoker

    def evaluate(
        self,
        request: policy_service_pb2.EvaluateAuthorizationRequest,
        *,
        options: CallOptions | None = None,
    ) -> authorization_decision_pb2.AuthorizationDecision:
        materialized, call = _prepare_evaluation(self._invoker, request, options)
        response = cast(
            policy_service_pb2.EvaluateAuthorizationResponse,
            self._invoker.unary(EVALUATE_AUTHORIZATION, materialized, call=call, retry_safe=True),
        )
        return required_response_message(
            response,
            "decision",
            authorization_decision_pb2.AuthorizationDecision,
            label="authorization evaluation",
        )

    def create(
        self,
        request: policy_service_pb2.CreateUsePolicyRequest,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        materialized = copy.deepcopy(request)
        if not materialized.HasField("use_policy") or not materialized.use_policy_id.strip():
            raise ValueError("policy create requires a generated policy and policy ID")
        parent = _project(self._invoker)
        if materialized.parent not in ("", parent):
            raise ValueError("policy parent must match the configured project")
        materialized.parent = parent
        materialized, call = _prepare_mutation(self._invoker, materialized, options)
        return _operation(
            self._invoker.unary(CREATE_USE_POLICY, materialized, call=call, retry_safe=True),
            "policy create",
        )

    def update(
        self,
        request: policy_service_pb2.UpdateUsePolicyRequest,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        if (
            not request.HasField("use_policy")
            or not request.HasField("update_mask")
            or not request.etag.strip()
        ):
            raise ValueError("policy update requires a policy, field mask, and etag")
        _policy_name(self._invoker, request.use_policy.name)
        materialized, call = _prepare_mutation(self._invoker, request, options)
        return _operation(
            self._invoker.unary(UPDATE_USE_POLICY, materialized, call=call, retry_safe=True),
            "policy update",
        )

    def get(
        self, name: str, *, if_none_match: str = "", options: CallOptions | None = None
    ) -> use_policy_pb2.UsePolicy:
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        response = cast(
            policy_service_pb2.GetUsePolicyResponse,
            self._invoker.unary(
                GET_USE_POLICY,
                policy_service_pb2.GetUsePolicyRequest(
                    name=_policy_name(self._invoker, name), if_none_match=if_none_match.strip()
                ),
                call=call,
                retry_safe=True,
            ),
        )
        return required_response_message(
            response, "use_policy", use_policy_pb2.UsePolicy, label="policy get"
        )

    def list(
        self,
        request: policy_service_pb2.ListUsePoliciesRequest | None = None,
        *,
        options: CallOptions | None = None,
    ) -> policy_service_pb2.ListUsePoliciesResponse:
        materialized = policy_service_pb2.ListUsePoliciesRequest()
        if request is not None:
            materialized.CopyFrom(request)
        parent = _project(self._invoker)
        if materialized.parent not in ("", parent):
            raise ValueError("policy list parent must match the configured project")
        materialized.parent = parent
        if materialized.HasField("page") and materialized.page.page_size > 1000:
            raise ValueError("policy page size cannot exceed 1000")
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        return cast(
            policy_service_pb2.ListUsePoliciesResponse,
            self._invoker.unary(LIST_USE_POLICIES, materialized, call=call, retry_safe=True),
        )

    def activate(
        self, name: str, etag: str, *, options: CallOptions | None = None
    ) -> operation_pb2.Operation:
        request = policy_service_pb2.ActivateUsePolicyRequest(
            name=_policy_name(self._invoker, name), etag=required_text("etag", etag)
        )
        request, call = _prepare_mutation(self._invoker, request, options)
        return _operation(
            self._invoker.unary(ACTIVATE_USE_POLICY, request, call=call, retry_safe=True),
            "policy activation",
        )

    def revoke(
        self,
        name: str,
        etag: str,
        reason_code: str,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        request = policy_service_pb2.RevokeUsePolicyRequest(
            name=_policy_name(self._invoker, name),
            etag=required_text("etag", etag),
            reason_code=required_text("reason_code", reason_code),
        )
        request, call = _prepare_mutation(self._invoker, request, options)
        return _operation(
            self._invoker.unary(REVOKE_USE_POLICY, request, call=call, retry_safe=True),
            "policy revocation",
        )

    def resolve_snapshot(
        self,
        name: str,
        effective_time: Timestamp,
        *,
        options: CallOptions | None = None,
    ) -> policy_reference_pb2.PolicyReference:
        effective_time.ToDatetime()
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        response = cast(
            policy_service_pb2.ResolvePolicySnapshotResponse,
            self._invoker.unary(
                RESOLVE_POLICY_SNAPSHOT,
                policy_service_pb2.ResolvePolicySnapshotRequest(
                    name=_policy_name(self._invoker, name), effective_time=effective_time
                ),
                call=call,
                retry_safe=True,
            ),
        )
        return required_response_message(
            response,
            "policy_snapshot",
            policy_reference_pb2.PolicyReference,
            label="policy snapshot resolution",
        )


class AsyncPolicies:
    """Asyncio variant of the generated-type-only Policy API."""

    def __init__(self, invoker: AsyncInvoker) -> None:
        self._invoker = invoker

    async def evaluate(
        self,
        request: policy_service_pb2.EvaluateAuthorizationRequest,
        *,
        options: CallOptions | None = None,
    ) -> authorization_decision_pb2.AuthorizationDecision:
        materialized, call = _prepare_evaluation(self._invoker, request, options)
        response = cast(
            policy_service_pb2.EvaluateAuthorizationResponse,
            await self._invoker.unary(
                EVALUATE_AUTHORIZATION, materialized, call=call, retry_safe=True
            ),
        )
        return required_response_message(
            response,
            "decision",
            authorization_decision_pb2.AuthorizationDecision,
            label="authorization evaluation",
        )

    async def create(
        self,
        request: policy_service_pb2.CreateUsePolicyRequest,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        materialized = copy.deepcopy(request)
        if not materialized.HasField("use_policy") or not materialized.use_policy_id.strip():
            raise ValueError("policy create requires a generated policy and policy ID")
        parent = _project(self._invoker)
        if materialized.parent not in ("", parent):
            raise ValueError("policy parent must match the configured project")
        materialized.parent = parent
        materialized, call = _prepare_mutation(self._invoker, materialized, options)
        return _operation(
            await self._invoker.unary(CREATE_USE_POLICY, materialized, call=call, retry_safe=True),
            "policy create",
        )

    async def update(
        self,
        request: policy_service_pb2.UpdateUsePolicyRequest,
        *,
        options: CallOptions | None = None,
    ) -> operation_pb2.Operation:
        if (
            not request.HasField("use_policy")
            or not request.HasField("update_mask")
            or not request.etag.strip()
        ):
            raise ValueError("policy update requires a policy, field mask, and etag")
        _policy_name(self._invoker, request.use_policy.name)
        materialized, call = _prepare_mutation(self._invoker, request, options)
        return _operation(
            await self._invoker.unary(UPDATE_USE_POLICY, materialized, call=call, retry_safe=True),
            "policy update",
        )

    async def get(
        self, name: str, *, if_none_match: str = "", options: CallOptions | None = None
    ) -> use_policy_pb2.UsePolicy:
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        response = cast(
            policy_service_pb2.GetUsePolicyResponse,
            await self._invoker.unary(
                GET_USE_POLICY,
                policy_service_pb2.GetUsePolicyRequest(
                    name=_policy_name(self._invoker, name), if_none_match=if_none_match.strip()
                ),
                call=call,
                retry_safe=True,
            ),
        )
        return required_response_message(
            response, "use_policy", use_policy_pb2.UsePolicy, label="policy get"
        )

    async def list(
        self,
        request: policy_service_pb2.ListUsePoliciesRequest | None = None,
        *,
        options: CallOptions | None = None,
    ) -> policy_service_pb2.ListUsePoliciesResponse:
        materialized = policy_service_pb2.ListUsePoliciesRequest()
        if request is not None:
            materialized.CopyFrom(request)
        parent = _project(self._invoker)
        if materialized.parent not in ("", parent):
            raise ValueError("policy list parent must match the configured project")
        materialized.parent = parent
        if materialized.HasField("page") and materialized.page.page_size > 1000:
            raise ValueError("policy page size cannot exceed 1000")
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        return cast(
            policy_service_pb2.ListUsePoliciesResponse,
            await self._invoker.unary(LIST_USE_POLICIES, materialized, call=call, retry_safe=True),
        )

    async def activate(
        self, name: str, etag: str, *, options: CallOptions | None = None
    ) -> operation_pb2.Operation:
        request, call = _prepare_mutation(
            self._invoker,
            policy_service_pb2.ActivateUsePolicyRequest(
                name=_policy_name(self._invoker, name), etag=required_text("etag", etag)
            ),
            options,
        )
        return _operation(
            await self._invoker.unary(ACTIVATE_USE_POLICY, request, call=call, retry_safe=True),
            "policy activation",
        )

    async def revoke(
        self, name: str, etag: str, reason_code: str, *, options: CallOptions | None = None
    ) -> operation_pb2.Operation:
        request, call = _prepare_mutation(
            self._invoker,
            policy_service_pb2.RevokeUsePolicyRequest(
                name=_policy_name(self._invoker, name),
                etag=required_text("etag", etag),
                reason_code=required_text("reason_code", reason_code),
            ),
            options,
        )
        return _operation(
            await self._invoker.unary(REVOKE_USE_POLICY, request, call=call, retry_safe=True),
            "policy revocation",
        )

    async def resolve_snapshot(
        self, name: str, effective_time: Timestamp, *, options: CallOptions | None = None
    ) -> policy_reference_pb2.PolicyReference:
        effective_time.ToDatetime()
        call = prepare_call(
            options, default_timeout=self._invoker.config.default_timeout, require_idempotency=False
        )
        response = cast(
            policy_service_pb2.ResolvePolicySnapshotResponse,
            await self._invoker.unary(
                RESOLVE_POLICY_SNAPSHOT,
                policy_service_pb2.ResolvePolicySnapshotRequest(
                    name=_policy_name(self._invoker, name), effective_time=effective_time
                ),
                call=call,
                retry_safe=True,
            ),
        )
        return required_response_message(
            response,
            "policy_snapshot",
            policy_reference_pb2.PolicyReference,
            label="policy snapshot resolution",
        )
